### ISSUE TITLE
How device-local applications will interact with the network

### ISSUE THREAD

**Issue by pfraze** [ssbc/phoenix#340]
After hemming and hawing on this topic a bit, we've come to the conclusion that applications should maintain their own keypairs and feeds, and thus their own identities. They will prepare their messages internally (set headers, add signature) then send them to the local sbot instance, where the messages are stored centrally and possibly replicated out. (Related work: https://github.com/ssbc/scuttlebot/pull/134 and https://github.com/ssbc/ssb-client)

The advantage to this is that permissions are straight-forward. We don't have to grant write-access to the user's main feed; each app uses its own. We can also treat applications as a type of ssb user and publish links to their feed IDs.

This opens some new issues for us to address. 

 1. How will program feeds announce themselves as "application feeds?" What are the implications of being an "application feed?"
 2. How will scuttlebot confirm to the user that, indeed, this is the feed of a application on their device, and not some rando on the network that's faking it?
 3. How will application feeds be associated with a user's main feed? How will it protect against users falsely asserting ownership of the application?
 4. If the user has to sunset their main feed (eg due to compromise or corruption) how will application feeds be re-associated to the new main feed?
 5. How will the user announce their application feeds to the rest of the network? By "following" them? Does that make sense, semantically, or is there another action-word we should use?

**Comment by pfraze** [comment]
When a local app connects to sbot, it proves ownership of its keypair during auth, and sbot can detect that the connection is over the loopback interface. I'd suggest that sbot stores metadata about the feed which includes `isLocal`, which phoenix could retrieve over RPC. This would solve 2.

I suspect the safest way to associate app feeds with user feeds is with bi-directional links, perhaps using `type: contact` messages. The application would publish a contact msg pointing to the user feed, with some attributes along the lines of `alias: true, role: user`, and the user would publish a contact msg to the app with something like `alias: true, role: app`. (This means the application would need to know the local user's id.)

To re-associate an app feed, you'd publish a new bi-directional link. The app's current user would be whichever feed the app most recently linked to and received a link from. If the app publishes a link and goes unrequited (doesn't get a link in reverse) it's in a bad state and isn't associated to anybody.

In Phoenix, the implications of being an application feed would be:
 - An app feed's profile would be labeled "application" and link to the owner.
 - An app feed's name would be show its owner's name (eg `pfraze (bittodo)`)
 - A user feed's profile would list all of its app feeds.
 - The users list would show app feeds underneath their owning users.
 - New device-local app feeds would get a call-to-action to confirm it.

**Comment by kumavis** [comment]
@pfraze with ethereum apps we solve a lot of problems by using proxy contracts that have a mutable whitelist of senders they allow to forward transactions/actions/messages. A feed can be pointed at a new proxy if the original one is compromised, or a feed can be removed from a proxy's whitelist if the feed is bad. seems to decouple things at the cost of indirection.

I'm missing a lot of ssb context, so this may not solve anything here.

**Comment by dominictarr** [comment]
can of worms, opened.

For the record, I think that there is a use for both approaches. You need this if you want a "parody account" so it's a low hanging, but tasty fruit. This problem of how do we associate these feeds with this account is the same as multiple devices.

These seem about equally complex to me.

**Comment by pfraze** [comment]
Yeah, we were always going to need a way to associate feeds at some point. The worms are now at large.

I don't know what a proxy feed would accomplish here, but I may just not be seeing the whole picture.

**Comment by dominictarr** [comment]
@kumavis ethereum has feeds? how do they work?

**Comment by kumavis** [comment]
Ethereum does not have feeds per-se but people and 'smart contracts' / programs send transactions that move money around and call other programs. Ethereum mostly does not distinguish between human accounts and programs.

The ethereum blockchain describes a simple VM. A program can write to a sandboxed disk and can be called with arguments. Running and storing anything costs a small amount of ether, the native cryptoasset. A program can query the callee and behave depending on that. Lets say you have a program that is a bank account for a small bike shop. In one implementation: The bank account program will only let the SmallBikeShopEmployee proxy program interact with it. To interact with it from your personal account, you proxy through that contract which has you on the whitelist. The proxying is not anything built into Ethereum, the proxy account would be a program coded to check the callee on a whitelist and then create a transaction with the passed in arguments.

theres a lot going on to explain on an unrelated thread : P

Would love to share in detail as I've dug pretty deep into it by now.

**Comment by dominictarr** [comment]
@kumavis I have read about ethereum so I feel that was enough to clarify what sense you ment "proxy"

**Comment by dominictarr** [comment]
@pfraze hmm, okay I didn't notice your comment about `alias` the first time I read this. sorry.
Yes I think this is good. We have to have bidirectional links, else someone could pretend to be your phone... which might fool other people into messaging the wrong person.

**Comment by pfraze** [comment]
@dominictarr to create the alias, the application needs to publish a link to the local user feed. That means the application either needs to be able to fetch the user feed id, or we need a sbot method that publishes the alias link for the application (thus hiding the id).

All ssb data is currently public, and write-access is controlled by keys, so it seems logical that an application (even on the Web) could safely have open API access to sbot. (That's presuming the apis work as expected, and don't have security bugs.)

However, if you read the local cache of messages you can get *some* idea of who the local user is, or at least know what network the user is a part of. In fact, if you look at the first message in the log index (which is ordered by local receive time) then there's a good chance you'll see the init message of the local main feed. Seems like it's an all-or-nothing situation, here are the choices I see:

 1. Try to hide the user's identity by sandboxing an app so it can only read its own messages. This is a complex solution for the same reason that encrypted feeds are complex right now, because the indexes aren't separated per-feed, so how do we keep data hidden from the indexes? Filter on read? Multiple indexes? But if this is something we *have* to solve, maybe now's the time to solve it.

Something else to consider, how much utility do applications lose if they can't read messages by other apps or by other users?

 2. Don't try to hide the user's identity, and give applications access to `whoami` (with challenge-signing to prove key ownership) so that the application can publish the association link. This is simpler and easier to do right now. For web apps, I'd want to have an authorization prompt before giving *any* api access, which I sort of hate but I think is necessary (otherwise SSB becomes a browsing tracker tool).

We can also consider starting with 2 and adding 1 later, eg changing it so an app defaults to 1 but can be granted permission for 2.

**Comment by pfraze** [comment]
After talking about this more with some folks, I'd suggest we go forward with 2 and not even use an auth prompt for now. So long as we have a path for securing it in the future, I'm ok with a more lax approach with better UX in the short term.

**Comment by pfraze** [comment]
Thinking out loud....

I've tried a few schemas for the alias link. I wasn't very satisfied with `alias: true, role: _` for two reasons: 
 - `role` didnt quite seem right;
 - it uses two attrs which work together, but according to `type: contact` behavior they could be emitted in two separate messages, which is more complex to process

Last night I settled on `myapp: true` and `myuser: true`, eg
```js
theapp emits { type: 'contact', contact: { feed: bob }, myuser: true }
bob emits { type: 'contact', contact: { feed: theapp }, myapp: true }
```
It's a problem that they can be combined though, eg what would it mean if:
```js
theapp emits { type: 'contact', contact: { feed: bob }, myuser: true, myapp: true }
```
so perhaps instead it should be `alias: type`, eg
```js
theapp emits { type: 'contact', contact: { feed: bob }, alias: 'user' }
bob emits { type: 'contact', contact: { feed: theapp }, alias: 'app' }
```
but `alias` seems unclear in that use, I'd almost prefer `ismy`
```js
theapp emits { type: 'contact', contact: { feed: bob }, ismy: 'user' }
bob emits { type: 'contact', contact: { feed: theapp }, ismy: 'app' }
```
which is colloquial but clear

semantics invite bikeshedding, but Im open to any strong opinions while I implement this. for reference, here's the existing attrs on contact messages:

```js
{
  type: 'contact',
  contact: { feed: $HASH },
  name: string,
  trust: -1|0|1,
  following: bool,
  profilePic: { ext: $HASH, type: string(mime), size: number, width: number, height: number }
}
```

**Comment by pfraze** [comment]
Propose for aliasing, app feed and main feed both publish a contact with
 - a "contact" link to the app feed
 - a "master" link to the main feed
eg:
```js
{ type: 'contact', contact: { feed: $app }, master: { feed: $main } }
```
App publishes it first about self, main feed publishes same message content to confirm

**Comment by dominictarr** [comment]
right, except that in the master feed $main === author, and in the alias feed $app == $author

**Comment by pfraze** [comment]
yeah. if an app publishes a new self-contact message with a different master, that should overwrite its old master. there's only 1 master at a time.

**Comment by dominictarr** [comment]
hmm, I think we can simplify this to just one link?
``` js
//the app publishes:
{type: 'contact', aka: {feed: primary (main feed) }, primary: true}
//the main feed publishes
{type: 'contact', aka: {feed: secondary }, primary: false, name: phone|photos|app|boat|server|etc}
```

would that be simpler?

**Comment by pfraze** [comment]
i like the regularity of always requiring a `contact` link to specify the subject.

I also prefer the `master` rel because it implies what `primary: bool` says, but does so without setting an additional attribute. a `contact` is computed as the LWW merge of all contact messages by one user about another user. a value that's contextualized by other values in the message complicates that. what would it mean to emit `primary: true/false` without an `aka` in the same message?

**Comment by dominictarr** [comment]
oh, yeah we should use `contact` instead of `aka`. That is more consistent with the other contact messages. though, these also have properties, like, `follows: true` or `trusts: 1/-1` if these all have properties, i don't see why delegation/aka/primary messages should be different. The only significant thing here is that a aka messages is invalid unless there it a matching reciprocal message.

**Comment by dominictarr** [comment]
but that is okay, because all of those message types have their own interpretation.

**Comment by pfraze** [comment]
the first thing I did was like you suggest, using the `contact` link and the attributes `myuser: true` (published by the app) and `myapp: true` (published by the user). I ditched that because they could combine (what if you publish `myapp: true, myuser: true`).

`primary: bool` isnt enough info, but we could do something like `alias: primary` from the app and `alias: application` from the user

**Comment by dominictarr** [comment]
but if there are only two states, well, that is how many settings a boolean has. I guess you have to check whether that boolean is != null instead of check whether it is truthy though.

**Comment by pfraze** [comment]
there are three states:
 - this contact is my primary alias
 - this contact is my secondary alias
 - this contact is not an alias to me

**Comment by dominictarr** [comment]
riiiight. okay that is true.

**Comment by pfraze** [comment]
ok, I'm going to implement the following scheme to replace the `master` link:

```js
//the app publishes:
{type: 'contact', contact: {feed: primary (main feed) }, alias: 'primary'}
//the main feed publishes
{type: 'contact', aka: {feed: secondary (app feed) }, alias: 'application'}
```

valid values for `alias`:
 - `'primary'`: this is the primary feed
 - `'application'`: this is an application used by the primary feed
 - falsey value: this contact is not my alias

**Comment by pfraze** [comment]
done
